File: packages/renderer-ios/ios/Sources/TerajsRendererHost/TerajsCommandApplier.swift
import UIKit

public enum UIKitHostApplyError: Error {
  case duplicateNode(Int)
  case expectedElement(Int)
  case expectedText(Int)
  case malformedCommand(UIKitHostCommandType)
  case missingNode(Int)
}

public final class UIKitCommandApplier {
  public private(set) var root: UIKitHostElementNode?

  private let viewFactory: UIKitHostViewFactory
  private var nodes: [Int: UIKitHostNode]

  public init(viewFactory: UIKitHostViewFactory = UIKitHostViewFactory()) {
    self.viewFactory = viewFactory
    self.nodes = [:]
  }

  public func applyCommands(_ commands: [UIKitHostCommand]) throws {
    for command in commands {
      try apply(command)
    }
  }

  public func apply(_ command: UIKitHostCommand) throws {
    switch command.type {
    case .createElement:
      try createElement(command)
    case .createText:
      try createText(command)
    case .insert:
      try insert(command)
    case .remove:
      try remove(command)
    case .setText:
      try setText(command)
    case .setProp:
      try setProp(command)
    case .setStyle:
      try setStyle(command)
    case .setClass:
      try setClass(command)
    case .subscribeEvent:
      try subscribeEvent(command)
    case .unsubscribeEvent:
      try unsubscribeEvent(command)
    }
  }

  public func node(for nodeId: Int) -> UIKitHostNode? {
    nodes[nodeId]
  }

  private func createElement(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId, let viewType = command.viewType else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }
    guard nodes[nodeId] == nil else {
      throw UIKitHostApplyError.duplicateNode(nodeId)
    }

    let node = UIKitHostElementNode(nodeId: nodeId, viewType: viewType, view: viewFactory.makeView(for: viewType))
    nodes[nodeId] = node
    root ??= node
  }

  private func createText(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId, let value = command.value?.stringValue ?? command.valueString else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }
    guard nodes[nodeId] == nil else {
      throw UIKitHostApplyError.duplicateNode(nodeId)
    }

    nodes[nodeId] = UIKitHostTextNode(nodeId: nodeId, value: value)
  }

  private func insert(_ command: UIKitHostCommand) throws {
    guard let parentId = command.parentId, let childId = command.childId else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    let parent = try requireElementNode(parentId)
    let child = try requireNode(childId)
    detachFromParent(child)

    let insertionIndex = resolvedInsertionIndex(in: parent.childNodeIds, anchorId: command.anchorId)
    parent.childNodeIds.insert(childId, at: insertionIndex)
    child.parentId = parentId

    if let element = child as? UIKitHostElementNode {
      attach(view: element.view, to: parent, anchorId: command.anchorId)
    } else if child is UIKitHostTextNode {
      UIKitHostViewUpdater.syncTextChildren(of: parent, allNodes: nodes)
    }
  }

  private func remove(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    let node = try requireNode(nodeId)
    if let parentId = node.parentId, let parent = nodes[parentId] as? UIKitHostElementNode {
      parent.childNodeIds.removeAll { $0 == node.nodeId }
      if let element = node as? UIKitHostElementNode {
        detach(view: element.view, from: parent.view)
      } else if node is UIKitHostTextNode {
        UIKitHostViewUpdater.syncTextChildren(of: parent, allNodes: nodes)
      }
    }

    removeSubtree(node)
    if root?.nodeId == nodeId {
      root = nil
    }
  }

  private func setText(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId, let value = command.value?.stringValue ?? command.valueString else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    let node = try requireTextNode(nodeId)
    node.value = value

    if let parentId = node.parentId, let parent = nodes[parentId] as? UIKitHostElementNode {
      UIKitHostViewUpdater.syncTextChildren(of: parent, allNodes: nodes)
    }
  }

  private func setProp(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId, let name = command.name else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    let node = try requireElementNode(nodeId)
    if let value = command.value, !value.isNullValue {
      node.props[name] = value
      UIKitHostViewUpdater.applyProp(named: name, value: value, to: node)
      return
    }

    node.props.removeValue(forKey: name)
    UIKitHostViewUpdater.applyProp(named: name, value: nil, to: node)
  }

  private func setStyle(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId, let style = command.style else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    let node = try requireElementNode(nodeId)
    for (name, value) in style {
      node.styles[name] = value
    }
    UIKitHostViewUpdater.applyStyles(node.styles, to: node)
  }

  private func setClass(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    let node = try requireElementNode(nodeId)
    node.className = command.className
    UIKitHostViewUpdater.applyClassName(command.className, to: node)
  }

  private func subscribeEvent(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId, let name = command.name else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    try requireElementNode(nodeId).subscribedEvents.insert(name)
  }

  private func unsubscribeEvent(_ command: UIKitHostCommand) throws {
    guard let nodeId = command.nodeId, let name = command.name else {
      throw UIKitHostApplyError.malformedCommand(command.type)
    }

    try requireElementNode(nodeId).subscribedEvents.remove(name)
  }

  private func requireElementNode(_ nodeId: Int) throws -> UIKitHostElementNode {
    guard let node = nodes[nodeId] else {
      throw UIKitHostApplyError.missingNode(nodeId)
    }
    guard let element = node as? UIKitHostElementNode else {
      throw UIKitHostApplyError.expectedElement(nodeId)
    }
    return element
  }

  private func requireNode(_ nodeId: Int) throws -> UIKitHostNode {
    guard let node = nodes[nodeId] else {
      throw UIKitHostApplyError.missingNode(nodeId)
    }
    return node
  }

  private func requireTextNode(_ nodeId: Int) throws -> UIKitHostTextNode {
    guard let node = nodes[nodeId] else {
      throw UIKitHostApplyError.missingNode(nodeId)
    }
    guard let text = node as? UIKitHostTextNode else {
      throw UIKitHostApplyError.expectedText(nodeId)
    }
    return text
  }

  private func detachFromParent(_ node: UIKitHostNode) {
    guard let parentId = node.parentId, let parent = nodes[parentId] as? UIKitHostElementNode else {
      return
    }

    parent.childNodeIds.removeAll { $0 == node.nodeId }
    if let element = node as? UIKitHostElementNode {
      detach(view: element.view, from: parent.view)
    } else if node is UIKitHostTextNode {
      UIKitHostViewUpdater.syncTextChildren(of: parent, allNodes: nodes)
    }
    node.parentId = nil
  }

  private func attach(view childView: UIView, to parent: UIKitHostElementNode, anchorId: Int?) {
    detach(view: childView, from: childView.superview)

    if let stackView = parent.view as? UIStackView {
      let index = arrangedSubviewIndex(in: stackView, anchorId: anchorId)
      stackView.insertArrangedSubview(childView, at: index)
      return
    }

    let index = subviewIndex(in: parent.view, anchorId: anchorId)
    parent.view.insertSubview(childView, at: index)
  }

  private func detach(view childView: UIView, from parentView: UIView?) {
    if let stackView = parentView as? UIStackView, stackView.arrangedSubviews.contains(childView) {
      stackView.removeArrangedSubview(childView)
    }
    childView.removeFromSuperview()
  }

  private func arrangedSubviewIndex(in stackView: UIStackView, anchorId: Int?) -> Int {
    guard let anchorId,
          let anchor = nodes[anchorId] as? UIKitHostElementNode,
          let index = stackView.arrangedSubviews.firstIndex(of: anchor.view) else {
      return stackView.arrangedSubviews.count
    }

    return index
  }

  private func subviewIndex(in view: UIView, anchorId: Int?) -> Int {
    guard let anchorId,
          let anchor = nodes[anchorId] as? UIKitHostElementNode,
          let index = view.subviews.firstIndex(of: anchor.view) else {
      return view.subviews.count
    }

    return index
  }

  private func resolvedInsertionIndex(in childNodeIds: [Int], anchorId: Int?) -> Int {
    guard let anchorId, let index = childNodeIds.firstIndex(of: anchorId) else {
      return childNodeIds.count
    }

    return index
  }

  private func removeSubtree(_ node: UIKitHostNode) {
    if let element = node as? UIKitHostElementNode {
      let childIds = element.childNodeIds
      for childId in childIds {
        if let child = nodes[childId] {
          removeSubtree(child)
        }
      }
      element.childNodeIds.removeAll()
      element.view.removeFromSuperview()
    }

    nodes.removeValue(forKey: node.nodeId)
  }
}

private extension UIKitHostCommand {
  var valueString: String? {
    if case let .string(value)? = value {
      return value
    }
    return nil
  }
}

File: packages/renderer-android/android/src/main/kotlin/dev/terajs/renderer/android/TerajsCommandApplier.kt
package dev.terajs.renderer.android

import android.content.Context
import android.view.View
import android.view.ViewGroup

class AndroidHostApplyException(message: String) : Exception(message)

class AndroidCommandApplier(context: Context, private val viewFactory: AndroidHostViewFactory = AndroidHostViewFactory(context)) {
  var root: AndroidHostElementNode? = null
    private set

  private val nodes = linkedMapOf<Int, AndroidHostNode>()

  fun applyCommands(commands: List<AndroidHostCommand>) {
    commands.forEach { command -> apply(command) }
  }

  fun apply(command: AndroidHostCommand) {
    when (command.type) {
      AndroidHostCommandType.CreateElement -> createElement(command)
      AndroidHostCommandType.CreateText -> createText(command)
      AndroidHostCommandType.Insert -> insert(command)
      AndroidHostCommandType.Remove -> remove(command)
      AndroidHostCommandType.SetText -> setText(command)
      AndroidHostCommandType.SetProp -> setProp(command)
      AndroidHostCommandType.SetStyle -> setStyle(command)
      AndroidHostCommandType.SetClass -> setClass(command)
      AndroidHostCommandType.SubscribeEvent -> subscribeEvent(command)
      AndroidHostCommandType.UnsubscribeEvent -> unsubscribeEvent(command)
    }
  }

  fun node(nodeId: Int): AndroidHostNode? = nodes[nodeId]

  private fun createElement(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val viewType = command.viewType ?: throw AndroidHostApplyException("${command.type} requires viewType")
    require(nodes[nodeId] == null) { "Duplicate Android host node $nodeId" }

    val node = AndroidHostElementNode(nodeId = nodeId, viewType = viewType, view = viewFactory.makeView(viewType))
    nodes[nodeId] = node
    if (root == null) {
      root = node
    }
  }

  private fun createText(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val value = command.value.stringValue ?: throw AndroidHostApplyException("${command.type} requires string value")
    require(nodes[nodeId] == null) { "Duplicate Android host node $nodeId" }

    nodes[nodeId] = AndroidHostTextNode(nodeId = nodeId, value = value)
  }

  private fun insert(command: AndroidHostCommand) {
    val parentId = command.parentId ?: throw AndroidHostApplyException("${command.type} requires parentId")
    val childId = command.childId ?: throw AndroidHostApplyException("${command.type} requires childId")
    val parent = requireElementNode(parentId)
    val child = requireNode(childId)

    detachFromParent(child)

    val index = resolvedInsertionIndex(parent.childNodeIds, command.anchorId)
    parent.childNodeIds.add(index, childId)
    child.parentId = parentId

    if (child is AndroidHostElementNode) {
      attach(child.view, parent, command.anchorId)
    } else if (child is AndroidHostTextNode) {
      AndroidHostViewUpdater.syncTextChildren(parent, nodes)
    }
  }

  private fun remove(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val node = requireNode(nodeId)

    node.parentId?.let { parentId ->
      val parent = nodes[parentId] as? AndroidHostElementNode
      if (parent != null) {
        parent.childNodeIds.remove(node.nodeId)
        if (node is AndroidHostElementNode) {
          detach(node.view, parent.view)
        } else if (node is AndroidHostTextNode) {
          AndroidHostViewUpdater.syncTextChildren(parent, nodes)
        }
      }
    }

    removeSubtree(node)
    if (root?.nodeId == nodeId) {
      root = null
    }
  }

  private fun setText(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val value = command.value.stringValue ?: throw AndroidHostApplyException("${command.type} requires string value")
    val node = requireTextNode(nodeId)
    node.value = value

    node.parentId?.let { parentId ->
      val parent = nodes[parentId] as? AndroidHostElementNode
      if (parent != null) {
        AndroidHostViewUpdater.syncTextChildren(parent, nodes)
      }
    }
  }

  private fun setProp(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val name = command.name ?: throw AndroidHostApplyException("${command.type} requires name")
    val node = requireElementNode(nodeId)

    if (command.value != null && !command.value.isNullValue) {
      node.props[name] = command.value
      AndroidHostViewUpdater.applyProp(name, command.value, node)
      return
    }

    node.props.remove(name)
    AndroidHostViewUpdater.applyProp(name, null, node)
  }

  private fun setStyle(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val style = command.style ?: throw AndroidHostApplyException("${command.type} requires style")
    val node = requireElementNode(nodeId)

    style.forEach { (name, value) -> node.styles[name] = value }
    AndroidHostViewUpdater.applyStyles(node.styles, node)
  }

  private fun setClass(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val node = requireElementNode(nodeId)
    node.className = command.className
    AndroidHostViewUpdater.applyClassName(command.className, node)
  }

  private fun subscribeEvent(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val name = command.name ?: throw AndroidHostApplyException("${command.type} requires name")
    requireElementNode(nodeId).subscribedEvents.add(name)
  }

  private fun unsubscribeEvent(command: AndroidHostCommand) {
    val nodeId = command.nodeId ?: throw AndroidHostApplyException("${command.type} requires nodeId")
    val name = command.name ?: throw AndroidHostApplyException("${command.type} requires name")
    requireElementNode(nodeId).subscribedEvents.remove(name)
  }

  private fun requireElementNode(nodeId: Int): AndroidHostElementNode {
    val node = nodes[nodeId] ?: throw AndroidHostApplyException("Missing Android host node $nodeId")
    return node as? AndroidHostElementNode
      ?: throw AndroidHostApplyException("Expected Android host element node $nodeId")
  }

  private fun requireNode(nodeId: Int): AndroidHostNode {
    return nodes[nodeId] ?: throw AndroidHostApplyException("Missing Android host node $nodeId")
  }

  private fun requireTextNode(nodeId: Int): AndroidHostTextNode {
    val node = nodes[nodeId] ?: throw AndroidHostApplyException("Missing Android host node $nodeId")
    return node as? AndroidHostTextNode
      ?: throw AndroidHostApplyException("Expected Android host text node $nodeId")
  }

  private fun detachFromParent(node: AndroidHostNode) {
    val parentId = node.parentId ?: return
    val parent = nodes[parentId] as? AndroidHostElementNode ?: return
    parent.childNodeIds.remove(node.nodeId)
    if (node is AndroidHostElementNode) {
      detach(node.view, parent.view)
    } else if (node is AndroidHostTextNode) {
      AndroidHostViewUpdater.syncTextChildren(parent, nodes)
    }
    node.parentId = null
  }

  private fun attach(childView: View, parent: AndroidHostElementNode, anchorId: Int?) {
    val parentView = parent.view as? ViewGroup ?: return
    detach(childView, childView.parent as? View)

    val index = if (anchorId != null) {
      val anchorNode = nodes[anchorId] as? AndroidHostElementNode
      val anchorView = anchorNode?.view
      if (anchorView != null) {
        val anchorIndex = parentView.indexOfChild(anchorView)
        if (anchorIndex >= 0) anchorIndex else parentView.childCount
      } else {
        parentView.childCount
      }
    } else {
      parentView.childCount
    }

    parentView.addView(childView, index)
  }

  private fun detach(childView: View, parentView: View?) {
    (parentView as? ViewGroup)?.removeView(childView)
  }

  private fun resolvedInsertionIndex(childNodeIds: List<Int>, anchorId: Int?): Int {
    if (anchorId == null) {
      return childNodeIds.size
    }
    val index = childNodeIds.indexOf(anchorId)
    return if (index >= 0) index else childNodeIds.size
  }

  private fun removeSubtree(node: AndroidHostNode) {
    if (node is AndroidHostElementNode) {
      val childIds = node.childNodeIds.toList()
      childIds.forEach { childId ->
        nodes[childId]?.let { child -> removeSubtree(child) }
      }
      node.childNodeIds.clear()
      (node.view.parent as? ViewGroup)?.removeView(node.view)
    }

    nodes.remove(node.nodeId)
  }
}

private val AndroidHostCommand.valueString: String?
  get() = (value as? TerajsJsonString)?.value